// Copyright AGNTCY Contributors (https://github.com/agntcy)
// SPDX-License-Identifier: Apache-2.0

syntax = "proto3";

package search.v1;

// A query to match the record against during discovery.
// For example:
//   Exact match:      { type: RECORD_QUERY_TYPE_NAME, value: "my-agent" }
//   Wildcard match:   { type: RECORD_QUERY_TYPE_NAME, value: "web*" }
//   Pattern match:    { type: RECORD_QUERY_TYPE_SKILL_NAME, value: "*machine*learning*" }
//   Question mark:    { type: RECORD_QUERY_TYPE_VERSION, value: "v1.0.?" }
//   List wildcards:   { type: RECORD_QUERY_TYPE_NAME, value: "agent-[0-9]" }
//   Complex match:    { type: RECORD_QUERY_TYPE_LOCATOR, value: "docker-image:https://*.example.com/*" }
message RecordQuery {
  // The type of the query to match against.
  RecordQueryType type = 1;

  // The query value to match against.
  // Supports wildcard patterns:
  //   '*' - matches zero or more characters
  //   '?' - matches exactly one character  
  //   '[]' - matches any character within brackets (e.g., [0-9], [a-z], [abc])
  string value = 2;
}

// Defines a list of supported record query types.
enum RecordQueryType {
  // Unspecified query type.
  RECORD_QUERY_TYPE_UNSPECIFIED = 0;

  // Query for a record name.
  // Supports wildcard patterns: "web*", "*service", "api-*-v2", "???api", "agent-[0-9]"
  RECORD_QUERY_TYPE_NAME = 1;

  // Query for a record version.
  // Supports wildcard patterns: "v1.*", "v2.*", "*-beta", "v1.0.?", "v[0-9].*"
  RECORD_QUERY_TYPE_VERSION = 2;

  // Query for a skill ID.
  // Numeric field - exact match only, no wildcard support.
  RECORD_QUERY_TYPE_SKILL_ID = 3;

  // Query for a skill name.
  // Supports wildcard patterns: "python*", "*script", "*machine*learning*", "Pytho?", "[A-M]*"
  RECORD_QUERY_TYPE_SKILL_NAME = 4;

  // Query for a locator type.
  // Supports wildcard patterns: "http*", "ftp*", "*docker*", "[hf]tt[ps]*"
  RECORD_QUERY_TYPE_LOCATOR = 5;

  // Query for an extension.
  // Supports wildcard patterns: "*-plugin", "*-extension", "core*", "ext-?", "plugin-[0-9]"
  RECORD_QUERY_TYPE_EXTENSION = 6;
}
